From: Andreas Gruenbacher <agruen@suse.de>
Subject: Fix compiling /usr/src/linux directly with O=
References: 65891

/etc/init.d/running-kernel creates the /usr/src/linux/include/asm
symlink among other things so that user-space programs which include
kernel headers directly won't fail miserably. While including kernel
headers directly is not recommended, it's sometimes the most sensible
thing to do. Now when trying to compile the /usr/src/linux tree with
a different output directory as below, a check in the main kernel
Makefile aborts the build. This patch fixes it.

  /usr/src/linux$ mkdir /var/tmp/obj
  /usr/src/linux$ cp arch/i386/defconfig.default /var/tmp/obj/.config
  /usr/src/linux$ make O=/var/tmp/obj

Signed-off-by: Andreas Gruenbacher <agruen@suse.de>

Index: linux-2.6.15/Makefile
===================================================================
--- linux-2.6.15.orig/Makefile
+++ linux-2.6.15/Makefile
@@ -301,7 +301,7 @@ AFLAGS_KERNEL	=
 # Needed to be compatible with the O= option
 LINUXINCLUDE    := -Iinclude \
                    $(if $(KBUILD_SRC),-Iinclude2 -I$(srctree)/include) \
-		   -include include/linux/autoconf.h
+		   -include $(objtree)/include/linux/autoconf.h
 
 CPPFLAGS        := -D__KERNEL__ $(LINUXINCLUDE)
 
@@ -430,7 +430,7 @@ ifeq ($(KBUILD_EXTMOD),)
 scripts: scripts_basic include/config/MARKER
 	$(Q)$(MAKE) $(build)=$(@)
 
-scripts_basic: include/linux/autoconf.h
+scripts_basic: $(objtree)/include/linux/autoconf.h
 
 # Objects we will link into vmlinux / subdirs we need to visit
 init-y		:= init/
@@ -458,12 +458,12 @@ include .config
 # with it and forgot to run make oldconfig.
 # If kconfig.d is missing then we are probarly in a cleaned tree so
 # we execute the config step to be sure to catch updated Kconfig files
-include/linux/autoconf.h: .kconfig.d .config
+$(objtree)/include/linux/autoconf.h: .kconfig.d .config
 	$(Q)mkdir -p include/linux
 	$(Q)$(MAKE) -f $(srctree)/Makefile silentoldconfig
 else
 # Dummy target needed, because used as prerequisite
-include/linux/autoconf.h: ;
+$(objtree)/include/linux/autoconf.h: ;
 endif
 
 # The all: target is the default when no target is given on the
@@ -804,7 +804,7 @@ endif
 # prepare2 creates a makefile if using a separate output directory
 prepare2: prepare3 outputmakefile
 
-prepare1: prepare2 include/linux/version.h include/asm \
+prepare1: prepare2 $(objtree)/include/linux/version.h $(objtree)/include/asm \
                    include/config/MARKER
 ifneq ($(KBUILD_MODULES),)
 	$(Q)rm -rf $(MODVERDIR)
@@ -828,14 +828,14 @@ export CPPFLAGS_vmlinux.lds += -P -C -U$
 #	hard to detect, but I suppose "make mrproper" is a good idea
 #	before switching between archs anyway.
 
-include/asm:
+$(objtree)/include/asm:
 	@echo '  SYMLINK $@ -> include/asm-$(ARCH)'
 	$(Q)if [ ! -d include ]; then mkdir -p include; fi;
 	@ln -fsn asm-$(ARCH) $@
 
 # 	Split autoconf.h into include/linux/config/*
 
-include/config/MARKER: scripts/basic/split-include include/linux/autoconf.h
+include/config/MARKER: scripts/basic/split-include $(objtree)/include/linux/autoconf.h
 	@echo '  SPLIT   include/linux/autoconf.h -> include/config/*'
 	@scripts/basic/split-include include/linux/autoconf.h include/config
 	@touch $@
@@ -859,7 +859,7 @@ define filechk_version.h
 	)
 endef
 
-include/linux/version.h: $(srctree)/Makefile .config FORCE
+$(objtree)/include/linux/version.h: $(srctree)/Makefile .config FORCE
 	$(call filechk,version.h)
 
 # ---------------------------------------------------------------------------
@@ -953,7 +953,7 @@ CLEAN_FILES +=	vmlinux System.map \
 
 # Directories & files removed with 'make mrproper'
 MRPROPER_DIRS  += include/config include2
-MRPROPER_FILES += .config .config.old include/asm .version .old_version \
+MRPROPER_FILES += .config .config.old $(objtree)/include/asm .version .old_version \
                   include/linux/autoconf.h include/linux/version.h \
 		  .kernelrelease Module.symvers tags TAGS cscope*
 
Index: linux-2.6.15/arch/arm/Makefile
===================================================================
--- linux-2.6.15.orig/arch/arm/Makefile
+++ linux-2.6.15/arch/arm/Makefile
@@ -177,7 +177,7 @@ endif
 archprepare: maketools
 
 .PHONY: maketools FORCE
-maketools: include/linux/version.h include/asm-arm/.arch FORCE
+maketools: $(objtree)/include/linux/version.h include/asm-arm/.arch FORCE
 	$(Q)$(MAKE) $(build)=arch/arm/tools include/asm-arm/mach-types.h
 
 # Convert bzImage to zImage
Index: linux-2.6.15/arch/powerpc/Makefile
===================================================================
--- linux-2.6.15.orig/arch/powerpc/Makefile
+++ linux-2.6.15/arch/powerpc/Makefile
@@ -176,7 +176,7 @@ archprepare: checkbin
 
 ifeq ($(CONFIG_PPC32),y)
 # Temporary hack until we have migrated to asm-powerpc
-include/asm: arch/$(ARCH)/include/asm
+$(objtree)/include/asm: arch/$(ARCH)/include/asm
 arch/$(ARCH)/include/asm: FORCE
 	$(Q)if [ ! -d arch/$(ARCH)/include ]; then mkdir -p arch/$(ARCH)/include; fi
 	$(Q)ln -fsn $(srctree)/include/asm-$(OLDARCH) arch/$(ARCH)/include/asm
Index: linux-2.6.15/arch/ppc/Makefile
===================================================================
--- linux-2.6.15.orig/arch/ppc/Makefile
+++ linux-2.6.15/arch/ppc/Makefile
@@ -117,7 +117,7 @@ archclean:
 archprepare: checkbin
 
 # Temporary hack until we have migrated to asm-powerpc
-include/asm: arch/$(ARCH)/include/asm
+$(objtree)/include/asm: arch/$(ARCH)/include/asm
 arch/$(ARCH)/include/asm:
 	$(Q)if [ ! -d arch/$(ARCH)/include ]; then mkdir -p arch/$(ARCH)/include; fi
 	$(Q)ln -fsn $(srctree)/include/asm-powerpc arch/$(ARCH)/include/asm
Index: linux-2.6.15/arch/sh/Makefile
===================================================================
--- linux-2.6.15.orig/arch/sh/Makefile
+++ linux-2.6.15/arch/sh/Makefile
@@ -138,7 +138,7 @@ boot := arch/sh/boot
 CPPFLAGS_vmlinux.lds := -traditional
 
 ifneq ($(KBUILD_SRC),)
-incdir-prefix	:= $(srctree)/include/asm-sh/
+	incdir-prefix	:= $(srctree)/include/asm-sh/
 else
 incdir-prefix	:=
 endif
@@ -173,7 +173,7 @@ include/asm-sh/.mach: $(wildcard include
 archprepare: maketools include/asm-sh/.cpu include/asm-sh/.mach
 
 .PHONY: maketools FORCE
-maketools:  include/linux/version.h FORCE
+maketools:  $(objtree)/include/linux/version.h FORCE
 	$(Q)$(MAKE) $(build)=arch/sh/tools include/asm-sh/machtypes.h
 
 all: zImage
Index: linux-2.6.15/scripts/Makefile.lib
===================================================================
--- linux-2.6.15.orig/scripts/Makefile.lib
+++ linux-2.6.15/scripts/Makefile.lib
@@ -100,7 +100,7 @@ __cpp_flags     = $(_cpp_flags)
 else
 
 # Prefix -I with $(srctree) if it is not an absolute path
-addtree = $(if $(filter-out -I/%,$(1)),$(patsubst -I%,-I$(srctree)/%,$(1))) $(1)
+addtree = $(1) $(if $(filter-out -I/%,$(1)),$(patsubst -I%,-I$(srctree)/%,$(1)))
 # Find all -I options and call addtree
 flags = $(foreach o,$($(1)),$(if $(filter -I%,$(o)),$(call addtree,$(o)),$(o)))
 
